lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py

```diff
@@ -561,145 +614,104 @@ def ensure_route_table_present(connection, module):
-                error_traceback=traceback.format_exc()
-            )
+    if subnets is not None:
+        associated_subnets = find_subnets(connection, module, vpc_id, subnets)
 
-        try:
-            result = ensure_subnet_associations(connection, vpc_id, route_table,
-                                                associated_subnets,
-                                                module.check_mode,
-                                                purge_subnets)
-            changed = changed or result['changed']
-        except EC2ResponseError as e:
-            raise AnsibleRouteTableException(
-                message='Unable to associate subnets for route table {0}, error: {1}'
-                .format(route_table, e),
-                error_traceback=traceback.format_exc()
-            )
+        result = ensure_subnet_associations(connection, module, route_table,
+                                            associated_subnets,
+                                            module.check_mode,
+                                            purge_subnets)
+        changed = changed or result['changed']
 
-    module.exit_json(changed=changed, route_table=get_route_table_info(route_table))
+    module.exit_json(changed=changed, route_table=get_route_table_info(connection, module, route_table))
 
 
 def main():
     argument_spec = ec2_argument_spec()
     argument_spec.update(
         dict(
-            lookup=dict(default='tag', required=False, choices=['tag', 'id']),
-            propagating_vgw_ids=dict(default=None, required=False, type='list'),
+            lookup=dict(default='tag', choices=['tag', 'id']),
+            propagating_vgw_ids=dict(type='list'),
             purge_routes=dict(default=True, type='bool'),
             purge_subnets=dict(default=True, type='bool'),
-            route_table_id=dict(default=None, required=False),
-            routes=dict(default=[], required=False, type='list'),
+            purge_tags=dict(default=False, type='bool'),
+            route_table_id=dict(),
+            routes=dict(default=[], type='list'),
             state=dict(default='present', choices=['present', 'absent']),
-            subnets=dict(default=None, required=False, type='list'),
-            tags=dict(default=None, required=False, type='dict', aliases=['resource_tags']),
-            vpc_id=dict(default=None, required=True)
+            subnets=dict(type='list'),
+            tags=dict(type='dict', aliases=['resource_tags']),
+            vpc_id=dict()
         )
     )
 
-    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+    module = AnsibleAWSModule(argument_spec=argument_spec,
+                              required_if=[['lookup', 'id', ['route_table_id']],
+                                           ['lookup', 'tag', ['vpc_id']],
+                                           ['state', 'present', ['vpc_id']]],
+                              supports_check_mode=True)
```

Comment:
It would be good to have some tests for check mode too.

Reply:
I'll give that a go when I revisit this for your other suggestions.

Reply:
It would be good but I'm really running out of steam on this PR.

Having said that, I noticed I've removed the DryRun handling added since I started this PR, which does a better job of checking permissions than the current approach